src/main/java/com/novocode/junit/JUnitRunner.java

```diff
@@ -1,201 +1,107 @@
-              Request request = Request.classes(cl);
-              if(globPatterns.size() > 0) {
-                request = new SilentFilterRequest(request, new GlobFilter(settings, globPatterns));
-              }
-              if(testFilter.length() > 0) {
-                request = new SilentFilterRequest(request, new TestFilter(testFilter, ed));
-              }
-              if(!includeCategories.isEmpty() || !excludeCategories.isEmpty()) {
-                request = new SilentFilterRequest(request,
-                        CategoryFilter.categoryFilter(true, loadClasses(testClassLoader, includeCategories), true,
-                                loadClasses(testClassLoader, excludeCategories)));
-              }
-              ju.run(request);
-            }
-          } catch(Exception ex) {
-            ed.testExecutionFailed(testClassName, ex);
-          }
-        } finally {
-          synchronized(System.getProperties()) {
-            for(Map.Entry<String, Object> me : oldprops.entrySet()) {
-              if(me.getValue() == NULL) {
-                System.clearProperty(me.getKey());
-              } else {
-                System.setProperty(me.getKey(), (String)me.getValue());
-              }
-            }
-          }
-        }
-        return new Task[0]; // junit tests do not nest
-      }
-    };
-  }
-
   private RunListener createRunListener(String runListenerClassName) {
-    try {
-      return (RunListener) testClassLoader.loadClass(runListenerClassName).newInstance();
-    } catch (Exception e) {
-      throw new RuntimeException(e);
-    }
-  }
-
-  private boolean shouldRun(Fingerprint fingerprint, Class<?> clazz, RunSettings settings) {
-    if(JUNIT_FP.equals(fingerprint)) {
-      // Ignore classes which are matched by the other fingerprints
-      if(TestCase.class.isAssignableFrom(clazz)) {
-        return false;
-      }
-      for(Annotation a : clazz.getDeclaredAnnotations()) {
-        if(a.annotationType().equals(RunWith.class)) return false;
+    if(runListenerClassName != null) {
+      try {
+        return (RunListener) testClassLoader.loadClass(runListenerClassName).newInstance();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
       }
-      return true;
-    } else {
-      RunWith rw = clazz.getAnnotation(RunWith.class);
-      if(rw != null) {
-        return !settings.ignoreRunner(rw.value().getName());
-      }
-      else return true;
-    }
-  }
-
-  private static Set<Class<?>> loadClasses(ClassLoader classLoader, Set<String> classNames) throws ClassNotFoundException {
-    Set<Class<?>> classes = new HashSet<Class<?>>();
-    for(String className : classNames) {
-      classes.add(classLoader.loadClass(className));
-    }
-    return classes;
+    } else return null;
   }
 
   @Override
   public String done() {
-    return "";
+    // Can't simply return the summary due to https://github.com/sbt/sbt/issues/3510
+    if(!used) return "";
```

Comment:
It'd be nice to use multi-line curly braces:
```
if (!used) {
  return "";
}
```

Reply:
This just looks convoluted to me. I know many people like this style but I have no idea why.